### PR TITLE
feat: add package to de-obfuscate Android stack traces

### PR DESCRIPTION
**Pull request by kruskall**
## Motivation/summary

Add package to parse and map android stack traces
Add tests

## Checklist

- [ ] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)
- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)
- [ ] Documentation has been updated

For functional changes, consider:
- Is it observable through the addition of either **logging** or **metrics**?
- Is its use being published in **telemetry** to enable product improvement?
- Have system tests been added to avoid regression?

## Related issues

Closes https://github.com/elastic/apm-server/issues/9644

### CODE REVIEW

File: internal/retrace/parse.go [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+type parser struct {
+	// used to keep track of the current class
+	// when descending into the method names.
+	currentClass string
+}
+
+// Mapping holds the information stored in the mapping file.
+type Mapping struct {
+	// obfuscated class name -> original class metdata
```
Comment (carsonip) [inline internal/retrace/parse.go:35]:
```suggestion
	// obfuscated class name -> original class metadata
```

File: internal/retrace/parse.go [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+		if err != nil {
+			return fmt.Errorf("failed to parse obfuscatedStartLine: %w", err)
+		}
+	}
+
+	originalStartLine := 0
+	// check if we have the original line numbers
+	originalMethodSignature, originalLineNumbers, ok := bytes.Cut(originalMethodInfo, []byte(":"))
+	if ok {
+		orgiginalStartLineStr, _, _ := bytes.Cut(bytes.TrimPrefix(originalLineNumbers, []byte(":")), []byte(":"))
```
Comment (carsonip) [inline internal/retrace/parse.go:186]:
```suggestion
		originalStartLineStr, _, _ := bytes.Cut(bytes.TrimPrefix(originalLineNumbers, []byte(":")), []byte(":"))
```

File: internal/retrace/parse.go [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+
+		var err error
+		originalStartLine, err = strconv.Atoi(string(orgiginalStartLineStr))
```
Comment (carsonip) [inline internal/retrace/parse.go:189]:
```suggestion
		originalStartLine, err = strconv.Atoi(string(originalStartLineStr))
```

**Review by carsonip — COMMENTED** [review]
still reading the PR but want to point out some minor typos first

**Review by simitt — COMMENTED** [review]
PR looks good. 

The testdata usecases are generally great, but I was wondering if the whole mapping file is required, or whether it could be reduced in size without losing value?
Ideally we could also use the mapping file to more easily verify that every part of the spec is implemented & tested. 
If that's not easily achievable, you could also look at adding a dedicated parser test, using a hand crafted, small mapping file that shows that the spec is fully implemented.

File: internal/retrace/map.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+	sb.WriteRune(' ')
+	sb.WriteString(originalClassName)
+	sb.WriteRune('.')
+	sb.WriteString(originalMethodNames[0].name)
+	sb.WriteRune('(')
+	sb.WriteString(sourceFile)
+	if hasLineNumber {
+		sb.WriteRune(':')
+		if originalMethodNames[0].line == 0 {
+			sb.WriteString(strconv.Itoa(obfuscatedLineNumber))
```
[the remark this answers is not in the document]
Comment (kruskall) [reply internal/retrace/map.go]: Done!

File: internal/retrace/map.go [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+	// trim spaces to avoid issues with editors
+	sourceFileInfo = strings.TrimSpace(sourceFileInfo)
+
+	// validate that the source file info is not malformed
+	if !strings.HasSuffix(sourceFileInfo, ")") {
+		return errors.New("missing ')' token suffix in source file info")
+	}
+
+	obfuscatedLineNumber := 0
```
[the remark this answers is not in the document]
Comment (kruskall) [reply internal/retrace/map.go:70]: The line number is part of the Stacktrace, that's a java object which is not part of the retrace specification. The equivalent document would be the JLS (Java Language Specification) or the jvm specification. AFAIK line numbers start from 1.

The following (keep it on the same/first line):

```java
public class TestLineNumber {public static void main(String args[]) {System.out.println("Line Number: "+Thread.currentThread().getStackTrace()[1].getLineNumber());}}
```

Should produce `Line Number: 1`

**Comment by kruskall** [comment]
> The testdata usecases are generally great, but I was wondering if the whole mapping file is required, or whether it could be reduced in size without losing value?

Good point! I've compressed them. Size went from 7Mb to 600K.

File: internal/retrace/map.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (kruskall) [reply internal/retrace/map.go]: > The line number is part of the Stacktrace, that's a java object which is not part of the retrace specification. The equivalent document would be the JLS (Java Language Specification) or the jvm specification. AFAIK line numbers start at 1.
> 
> The following (keep it on the same/first line):
> 
> ```java
> public class TestLineNumber {public static void main(String args[]) {System.out.println("Line Number: "+Thread.currentThread().getStackTrace()[1].getLineNumber());}}
> ```
> 
> Should produce `Line Number: 1`

It surely does. Unfortunately retrace does not care and uses zero line numbers just fine. :sob: 

I've update the code to use -1 (if that's a valid line number whoever designed the language is messing with us).

File: internal/retrace/parse.go [hunk 6]
```diff
@@ -199,6 +202,9 @@ func (p *parser) parseMethodline(m *Mapping, line []byte) error {
 	originalMethodName := originalMethodSignature[:argStart]
 
 	k := methodID{
+		// Always use startline in the key.
+		// If the line numbers are not present (either in the mapping or the stacktrace)
+		// a methoid with a line 0 is going to be used.
```
Comment (simitt) [inline internal/retrace/parse.go:210]:
```suggestion
		// a methodid with a line 0 is going to be used.
```

File: internal/retrace/parse.go [hunk 7]
```diff
@@ -193,6 +193,9 @@ func (p *parser) parseMethodline(m *Mapping, line []byte) error {
 		if err != nil {
 			return fmt.Errorf("failed to parse originalStartLine: %w", err)
 		}
+	} else {
+		// no inlined method, the leading line number are the original line numbers.
```
Comment (simitt) [inline internal/retrace/parse.go:197]:
```suggestion
		// no inlined method, the leading line numbers are the original line numbers.
```

**Review by simitt — APPROVED** [review]
Looks great, thanks! Only minor cosmetical suggestions.

**Pull request closed without merging**